### PR TITLE
add arm64 support for tinyinst

### PR DESCRIPTION
**Pull request by cl-g**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -88,8 +103,14 @@ if(${ARCHITECTURE} MATCHES arm64)
...
     common.h
     common.cpp
+    assembler.h
+    instruction.h
+    tinyinst.h
+    tinyinst.cpp
+    ${arch_specific_files}
     ${platform_specific_files}
   )
+  add_dependencies(tinyinst reil)
```
[the remark this answers is not in the document]
Comment (cl-g) [reply CMakeLists.txt]: done

File: arch/arm64/arm64_assembler.cpp [hunk 2]
```diff
@@ -0,0 +1,934 @@
...
+  uint32_t b_opc = b(0, 12);
+  tinyinst_.WriteCode(module, &b_opc, sizeof(b_opc));
+  tinyinst_.WriteCode(module, &value, sizeof(value));
+}
+
+void Arm64Assembler::FixOffset(ModuleInfo *module, uint32_t jmp_offset,
+                               uint32_t target_offset) {
+  uint64_t instr_address =
+      (uint64_t)module->instrumented_code_local + jmp_offset;
+  int32_t relative_offset = target_offset - jmp_offset;
```
[the remark this answers is not in the document]
Comment (cl-g) [reply arch/arm64/arm64_assembler.cpp:150]: I added a function to check for all that checks and encodes immediate based on signed/unsigned and the amount of available bits.

File: arch/arm64/arm64_assembler.cpp [hunk 3]
```diff
@@ -0,0 +1,934 @@
...
+    FATAL("offset out of range, max is [-0x1000, +0x1000], was: %d", offset);
+  }
+}
+
+void Arm64Assembler::InstrumentGlobalIndirect(ModuleInfo *module,
+                                              Instruction &inst,
+                                              size_t instruction_address) {
+  FATAL(
+      "-indirect_instrumentation=global is not supported on arm64 \
+         please use 'local'");
```
[the remark this answers is not in the document]
Comment (cl-g) [reply arch/arm64/arm64_assembler.cpp:263]: done, thx for the ptr

File: arch/arm64/arm64_helpers.cpp [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+  // LD*R/B/H == 0b00
+  return load_store(0b00, size, data_reg, base_reg, offset);
+}
+
+
+// TODO: change op to reil instruction constant
+uint32_t add_sub_reg_imm(uint8_t op, Register dst_reg, Register src_reg, uint32_t offset) {
+  uint32_t instr = 0;
+
+  if(offset > 0x1000) {
```
[the remark this answers is not in the document]
Comment (cl-g) [reply arch/arm64/arm64_helpers.cpp]: yeeeep, good catch - Encode(Un)SignedImmediate should fix that

File: arch/arm64/arm64_assembler.cpp [hunk 5]
```diff
@@ -0,0 +1,934 @@
...
+      addr += GetRipRelativeOffset(inst);
+      auto dst_reg = std::get<arm64::Register>(inst.instr.operands[0]);
+      EmitLoadLit(module, reg(dst_reg), 64, false, addr);
+      uint32_t ldr_instr = ldr(dst_reg.size, reg(dst_reg), reg(dst_reg), 0);
+      tinyinst_.WriteCode(module, &ldr_instr, sizeof(ldr_instr));
+      break;
+    }
+
+    case arm64::kLdrsLiteral: {
+      FATAL("arm64::kLdrsLiteral");
```
[the remark this answers is not in the document]
Comment (cl-g) [reply arch/arm64/arm64_assembler.cpp:555]: added to TODO list :)

File: instruction.h [hunk 6]
```diff
@@ -49,6 +51,7 @@ struct Instruction {
     bbend(false),
     iclass(InstructionClass::INVALID),
 #ifdef ARM64
+  instr({})
```
[the remark this answers is not in the document]
Comment (cl-g) [reply instruction.h]: done

File: macOS/debugger.cpp [hunk 7]
```diff
@@ -1651,6 +1650,8 @@ char **Debugger::GetEnvp() {
     p++;
   }
 
+  additional_env.push_back("DYLD_SHARED_REGION=private");
```
[the remark this answers is not in the document]
Comment (cl-g) [reply macOS/debugger.cpp]: agreed, adding as flag is much cleaner - thanks for the suggestion!

File: third_party/CMakeLists.txt [hunk 8]
```diff
@@ -56,3 +68,4 @@ add_custom_target(
   xed
   DEPENDS ${XED_LIB} ${XED_INTERFACE_H}
 )
+endif()
```
[the remark this answers is not in the document]
Comment (cl-g) [reply third_party/CMakeLists.txt]: done

File: tinyinst.cpp [hunk 9]
```diff
@@ -323,6 +316,7 @@ bool TinyInst::HandleIndirectJMPBreakpoint(void *address) {
   size_t list_head_offset;
   size_t instruction_address = 0;
 
+  IndirectBreakpoinInfo bp_info = {};
```
[the remark this answers is not in the document]
Comment (cl-g) [reply tinyinst.cpp:318]: yes, [init](https://en.cppreference.com/w/cpp/language/zero_initialization) is done to 0 (also double checked compiler output on different optimization levels)

File: tinyinst.h [hunk 10]
```diff
@@ -255,7 +251,7 @@ class ModuleInfo {
 
   // per callsite jumplist breakpoint
   // from breakpoint address to list head offset
-  std::unordered_map<size_t, TinyInst::IndirectBreakpoinInfo> br_indirect_newtarget_list;
+  std::unordered_map<size_t, IndirectBreakpoinInfo> br_indirect_newtarget_list;
```
[the remark this answers is not in the document]
Comment (cl-g) [reply tinyinst.h:262]: pure magic - I was about to check the chain of includes where one forward declares the struct but not sure where it actually comes from :D --> changed it

**Comment by ifratric** [comment]
Thanks, merging!

**Pull request merged**